src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/ImmutableDispatchRuntime.cs

```diff
@@ -640,6 +667,44 @@ internal async Task<MessageRpc> ProcessMessageAsync(MessageRpc rpc)
             return rpc;
         }
 
+        private (bool isConfigured, HttpContext httpContext, string authenticationScheme) IsCustomAuthenticationSchemeConfigured()
+            =>  OperationContext.Current.IncomingMessageProperties.TryGetValue<HttpContext>(
+                    "Microsoft.AspNetCore.Http.HttpContext", out var httpContext)
+                && httpContext.Items.TryGetValue("CoreWCF.Channels.HttpTransportSettings.CustomAuthenticationScheme",
+                    out var authenticationScheme)
+                && authenticationScheme is string authenticationSchemeValue
+                && !string.IsNullOrEmpty(authenticationSchemeValue)
+                ? (true, httpContext, authenticationSchemeValue)
+                : (false, null, null);
+
+        private async ValueTask<(bool isAuthorized, Message message)> EvaluateAuthorizationPolicies(MessageRpc rpc)
+        {
+            var httpContext = OperationContext.Current.IncomingMessageProperties["Microsoft.AspNetCore.Http.HttpContext"] as HttpContext;
+            string authenticationScheme = httpContext.Items["CoreWCF.Channels.HttpTransportSettings.CustomAuthenticationScheme"] as string;
+            var principal = httpContext.User;
+            AuthorizationOperationInvoker invoker = (AuthorizationOperationInvoker)rpc.Operation.Invoker;
+
+            if (!principal.Identity.IsAuthenticated)
+            {
+                if (!invoker.IsAllowAnonymous)
+                {
```

Comment:
There's a potential DoS problem with denying anonymous requests this late in the pipe. For a request to have got this far, there might have been a lot of resources allocated. For example, using WSHttpBinding you could have established a session context which would allocate quite a bit of bookkeeping objects which are expensive to create (creates a cryptographic random session key which is always expensive to create) as well as potentially instantiated a service instance (depending on instance context mode). If we allow anonymous requests through on an endpoint which is configured with authentication to support operations attributed as allowing anonymous users, a malicious attacker can cheaply send through a large number of anonymous requests which cause a lot of memory and CPU usage, potentially on a binding which is really expensive (like WSHttpBinding) and cause a service outage. A basic principal is do as little work as possible and allocate as few resources as possible until you have authenticated the user. If you want to allow anonymous users to make calls, then that's something you are explicitly opting in to and can evaluate the risk for the binding you are using. For example, allowing anonymous users when using a sessionless binding such as BasicHttpBinding, and using InstanceContextMode.Singleton, your exposure to a malicious client forcing resource allocation is minimal.  
Basically as I said in the other comment, anonymous operations are supported through a separate endpoint with an interface which only defines the methods allowed to be called anonymously.

Reply:
Supporting [AllowAnonymous] is indeed a bit challenging as we have to go deep into the pipeline to know which operation will be executed before rejecting the request. In a previous commit i was delaying the policy evaluation to the IOperationInvoker and then I realized we could be exposed to DoS attack. I tried to move the policy evaluation sooner in the pipeline but it's indeed still too late.

Do you think we should warn users about [AllowAnonymous] usages ?

Reply:
Do you think we should emit a warning log or should we throw and refuse to startup if it's used? I worry about discoverability if it's a warning, but it's also harmless in that they won't accidentally process an unauthenticated request if they miss the warning.

Reply:
Did this get resolved by dropping the Anonymous support?

Reply:
Yes. Non authenticated requests are now dropped earlier in the request pipeline.

Reply:
Regarding warning we may add a roslyn analyzer with code fix to detect:
- `[AllowAnonymous]` usages at operation contract level
- `[Authorize]` at service contract level

Reply:
Added to https://github.com/CoreWCF/CoreWCF/pull/850/commits/ce7e91b13a1204ae1565172975accb7386256fa6 to detect above usages